runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java

```diff
@@ -261,96 +275,123 @@ State waitUntilFinish(
     long startNanos = nanoClock.nanoTime();
 
-    State state;
+    State state = State.UNKNOWN;
+    Exception exception;
     do {
-      // Get the state of the job before listing messages. This ensures we always fetch job
-      // messages after the job finishes to ensure we have all them.
-      state =
-          getStateWithRetries(
-              BackOffAdapter.toGcpBackOff(STATUS_BACKOFF_FACTORY.withMaxRetries(0).backoff()),
-              sleeper);
-      boolean hasError = state == State.UNKNOWN;
-
-      if (messageHandler != null && !hasError) {
-        // Process all the job messages that have accumulated so far.
-        try {
-          List<JobMessage> allMessages = monitor.getJobMessages(getJobId(), lastTimestamp);
-
-          if (!allMessages.isEmpty()) {
-            lastTimestamp =
-                fromCloudTime(allMessages.get(allMessages.size() - 1).getTime()).getMillis();
-            messageHandler.process(allMessages);
-          }
-        } catch (GoogleJsonResponseException | SocketTimeoutException e) {
-          hasError = true;
-          LOG.warn("There were problems getting current job messages: {}.", e.getMessage());
-          LOG.debug("Exception information:", e);
-        }
+      exception = null;
+      try {
+        // Get the state of the job before listing messages. This ensures we always fetch job
+        // messages after the job finishes to ensure we have all them.
+        state =
+            getStateWithRetries(
+                BackOffAdapter.toGcpBackOff(STATUS_BACKOFF_FACTORY.withMaxRetries(0).backoff()),
+                sleeper);
+      } catch (IOException e) {
+        exception = e;
+        LOG.warn("Failed to get job state: {}", e.getMessage());
+        LOG.debug("Failed to get job state: {}", e);
+        continue;
       }
 
-      if (!hasError) {
-        // We can stop if the job is done.
-        if (state.isTerminal()) {
-          switch (state) {
-            case DONE:
-            case CANCELLED:
-              LOG.info("Job {} finished with status {}.", getJobId(), state);
-              break;
-            case UPDATED:
-              LOG.info(
-                  "Job {} has been updated and is running as the new job with id {}. "
-                      + "To access the updated job on the Dataflow monitoring console, "
-                      + "please navigate to {}",
-                  getJobId(),
-                  getReplacedByJob().getJobId(),
-                  MonitoringUtil.getJobMonitoringPageURL(
-                      getReplacedByJob().getProjectId(),
-                      getRegion(),
-                      getReplacedByJob().getJobId()));
-              break;
-            default:
-              LOG.info("Job {} failed with status {}.", getJobId(), state);
-          }
-          return state;
-        }
+      exception = processJobMessages(messageHandler, monitor);
+
+      if (exception != null) {
```

Comment:
previously we would reset backoff before `continue`, is it the right thing to not do this anymore?

Reply:
Previous code is actually too layered and that's why I did refactored it a lot. backoff.reset() is under if(!hasError) condition.

We did reset backoff only in case if there was no error, ie status is not UNKNOWN. That caused us to fail upon reaching max retry count while receiving UNKNOWN state, not upon reaching timeout.

My change separates two cases:
a) If we fail to get status, we do not reset backoff and will fail due to exceptions upon attempt to get job status
b) If we actually receive UNKNOWN status, we will wait until timeout.

Reply:
Let me try to summarize the main flow to see if I understand it correctly:

**Previous Flow**
* get job state:
  * get non-`UNKNOWN` state -> **reset backoff** -> continue loop if not terminal;
     * will timeout at max duration or get a terminal state; correct behavior;
  * get `IOException`, same as:
  * get `UNKNOWN` state -> continue loop unconditionally;
     * does not **not** reset backoff;
     * can exceed number of allowed attempts fast, not waiting for max allowed duration;

**New Flow**
* get job state:
  * get non-`UNKNOWN` or `UNKNOWN` state -> **reset backoff** -> continue loop if not terminal;
    * can only receive `UNKNOWN` explicitly;
    * will timeout at max duration or get a terminal state; correct behavior;
  * get `IOException` -> continue loop unconditionally:
    * can exceed number of attempts instead of waiting for max allowed time;

Reply:
In this case the logic seems right. I would probably try to organize the body of the loop to emphasize the flow though, something along the lines of:

```
Optional<State> state = tryGetState(); 
if (!state.isPresent() || !tryProcessJobMessages()) {
  continue;
}

if (state.get().isTerminal()) {
  return state.get();
}

resetAttemptsCount();
```

Hope this makes sense

Reply:
You are right and your code example makes sense. The problem that didn't let me to get that state is that I want to propagate exception to outside of the loop. Unfortunately, java can not pass method arguments by reference, so there's no clear way to return state or exception except defining explicit class. And that would be a bit of an overkill in this case.